pkg/storage/stores/shipper/util/cached_client.go

```diff
@@ -0,0 +1,111 @@
+// Cache items are evicted after first read or a timeout. The cache is rebuilt during List call with empty prefix or we encounter a cache miss.
+type CachedObjectClient struct {
+	chunk.ObjectClient
+	tables       map[string][]chunk.StorageObject
+	tablesMtx    sync.Mutex
+	cacheBuiltAt time.Time
+}
+
+func NewCachedObjectClient(downstreamClient chunk.ObjectClient) *CachedObjectClient {
+	return &CachedObjectClient{
+		ObjectClient: downstreamClient,
+		tables:       map[string][]chunk.StorageObject{},
+	}
+}
+
+func (c *CachedObjectClient) List(ctx context.Context, prefix, _ string) ([]chunk.StorageObject, []chunk.StorageCommonPrefix, error) {
+	c.tablesMtx.Lock()
+	defer c.tablesMtx.Unlock()
+
+	if prefix == "" {
+		tables, err := c.listTables(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return []chunk.StorageObject{}, tables, nil
+	}
+
+	// While listing objects in a table, prefix is set to <table-name>+delimiter so trim the delimiter first.
+	tableName := strings.TrimSuffix(prefix, delimiter)
+	if strings.Contains(tableName, delimiter) {
+		return nil, nil, fmt.Errorf("invalid prefix %s for listing table objects", prefix)
+	}
+	tableObjects, err := c.listTableObjects(ctx, tableName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return tableObjects, []chunk.StorageCommonPrefix{}, nil
+}
+
+// listTables assumes that tablesMtx is already locked by the caller
+func (c *CachedObjectClient) listTables(ctx context.Context) ([]chunk.StorageCommonPrefix, error) {
+	// do a flat listing by setting delimiter to empty string
+	objects, _, err := c.ObjectClient.List(ctx, "", "")
+	if err != nil {
+		return nil, err
+	}
+
+	// build the cache and response containing just table names as chunk.StorageCommonPrefix
+	var tableNames []chunk.StorageCommonPrefix
+	for _, object := range objects {
+		ss := strings.Split(object.Key, delimiter)
+		if len(ss) != 2 {
+			return nil, fmt.Errorf("invalid object key found %s", object.Key)
+		}
+
+		if _, ok := c.tables[ss[0]]; !ok {
+			tableNames = append(tableNames, chunk.StorageCommonPrefix(ss[0]))
+		}
+		c.tables[ss[0]] = append(c.tables[ss[0]], object)
+	}
+
+	c.cacheBuiltAt = time.Now()
+
+	return tableNames, nil
+}
+
+// listTableObjects assumes that tablesMtx is already locked by the caller
+func (c *CachedObjectClient) listTableObjects(ctx context.Context, tableName string) ([]chunk.StorageObject, error) {
+	objects, ok := c.tables[tableName]
+	if ok && c.cacheBuiltAt.Add(cacheTimeout).After(time.Now()) {
+		// evict the element read from cache
+		delete(c.tables, tableName)
+		return objects, nil
+	}
+
+	// requested element not found in the cache, rebuild the cache.
+	_, err := c.listTables(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	objects = c.tables[tableName]
+	// evict the element read from cache
+	delete(c.tables, tableName)
```

Comment:
why do we evict element read from cache ?

Reply:
This is to avoid any possible bugs if the caller misses refreshing the cache explicitly.
The cache is integrated to make it look seamless for consumers since it only works for hosted object stores.
Removing it should not be big of a problem since we have a ttl but I just added it as an extra protection mechanism.

Reply:
but if we ask twice the same table then we will call twice List on the objectstore ?

Reply:
yeah but we don't have a use case where we will list objects from the same table twice. We usually circle through a list of tables for sync or compaction so we will not hit this case most likely. The only case it can happen is if someone has a really low sync interval and the non-deterministic [iteration of maps](https://github.com/grafana/loki/blob/923671a17e250542e10037f514e8f5c21fa205b9/pkg/storage/stores/shipper/downloads/table_manager.go#L210) in go can make the same table to be synced back to back.

Reply:
We can always remove this if we see it causing a problem.